source/common/upstream/conn_pool_map.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <map>
+#include <vector>
+
+#include "envoy/event/dispatcher.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Upstream {
+/**
+ *  A class mapping keys to connection pools, with some recycling logic built in.
+ */
+template <typename KEY_TYPE, typename POOL_TYPE> class ConnPoolMap {
+public:
+  using PoolFactory = std::function<std::unique_ptr<POOL_TYPE>()>;
+  using DrainedCb = std::function<void()>;
+
+  ConnPoolMap(Event::Dispatcher& dispatcher);
+  ~ConnPoolMap();
+  /**
+   * Returns an existing pool for `key`, or creates a new one using `factory`. Note that it is
+   * possible for this to fail if a limit on the number of pools allowed is reached.
+   * @return The pool corresponding to `key`, or `absl::nullopt`
+   */
+  absl::optional<POOL_TYPE*> getPool(KEY_TYPE key, PoolFactory factory);
```

Comment:
nit: Does storing an std::reference_wrapper inside absl::optional work? If so can we do that? If not ignore.

Reply:
On additional read, it looks like in this PR this will always succeed. Can we just make this return a reference now and add the optional behavior in a later PR?

Reply:
Done